enterprise/cypher/slotted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/slotted/pipes/DistinctSlottedPrimitivePipe.scala

```diff
@@ -0,0 +1,92 @@
+import org.neo4j.cypher.internal.runtime.interpreted.ExecutionContext
+import org.neo4j.cypher.internal.runtime.interpreted.commands.expressions.Expression
+import org.neo4j.cypher.internal.runtime.interpreted.pipes.{Pipe, PipeWithSource, QueryState}
+import org.neo4j.cypher.internal.runtime.slotted.SlottedExecutionContext
+import org.neo4j.cypher.internal.runtime.slotted.helpers.SlottedPipeBuilderUtils
+import org.neo4j.cypher.internal.runtime.{LongArraySet, PrefetchingIterator}
+import org.neo4j.cypher.internal.util.v3_4.attribution.Id
+
+import scala.collection.immutable
+
+case class DistinctSlottedPrimitivePipe(source: Pipe,
+                                        slots: SlotConfiguration,
+                                        primitiveSlots: Array[Int],
+                                        projections: Map[Slot, Expression])
+                                       (val id: Id = Id.INVALID_ID)
+  extends PipeWithSource(source) {
+
+  //===========================================================================
+  // Compile-time initializations
+  //===========================================================================
+  private val setValuesInOutput: immutable.Iterable[(ExecutionContext, QueryState, ExecutionContext) => Unit] = projections.map {
+    case (slot, expression) =>
+      val f = SlottedPipeBuilderUtils.makeSetValueInSlotFunctionFor(slot)
+      (incomingContext: ExecutionContext, state: QueryState, outgoingContext: ExecutionContext) =>
+        f(outgoingContext, expression(incomingContext, state))
+  }
+
+  projections.values.foreach(_.registerOwningPipe(this))
+
+  //===========================================================================
+  // Runtime code
+  //===========================================================================
+  protected def internalCreateResults(input: Iterator[ExecutionContext],
+                                      state: QueryState): Iterator[ExecutionContext] = {
+    new PrefetchingIterator[ExecutionContext] {
+      private val seen = new LongArraySet(32, projections.size)
+
+      override def produceNext(): Option[ExecutionContext] = {
+        while (input.nonEmpty) {
+          val next: ExecutionContext = input.next()
+
+          // Create key array
+          val keys = buildKey(next)
+
+          if (seen.add(keys)) {
+            // Found something! Set it as the next element to yield, and exit
+            val outgoing = SlottedExecutionContext(slots)
+            for (setter <- setValuesInOutput) {
+              setter(next, state, outgoing)
+            }
+
+            return Some(outgoing)
+          }
+        }
+
+        None
+      }
+    }
+  }
+
+  private def buildKey(next: ExecutionContext): Array[Long] = {
+    val keys = new Array[Long](primitiveSlots.length)
+    var i = 0
+    while (i < primitiveSlots.length) {
+      keys(i) = next.getLongAt(primitiveSlots(i))
+      i += 1
+    }
```

Comment:
There is a special case of this pipe where the slots are consecutive in the execution context when we could use System.arraycopy here, but I don't know if it is worth optimizing for.